internal/controllers/app_controller.go

```diff
@@ -270,12 +314,298 @@ func (r *AppReconciler) reconcile(ctx context.Context, app *ketchv1.App, logger
+			timeoutEvent := newAppDeploymentEvent(app, appReconcileError, fmt.Sprintf("deployment timeout: %s", err.Error()))
+			recorder.AnnotatedEventf(app, timeoutEvent.Annotations, v1.EventTypeWarning, timeoutEvent.Reason, timeoutEvent.Description)
+			return err
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+
+		dep, err = cli.AppsV1().Deployments(namespace).Get(context.TODO(), dep.Name, metav1.GetOptions{})
+		if err != nil {
+			deploymentErrorEvent := newAppDeploymentEvent(app, appReconcileError, fmt.Sprintf("error getting deployments: %s", err.Error()))
+			recorder.AnnotatedEventf(app, deploymentErrorEvent.Annotations, v1.EventTypeWarning, deploymentErrorEvent.Reason, deploymentErrorEvent.Description)
+			return err
+		}
+	}
+
+	outcome := ketchv1.AppReconcileOutcome{AppName: app.Name, DeploymentCount: int(dep.Status.ReadyReplicas)}
+	outcomeEvent := newAppDeploymentEvent(app, appReconcileComplete, outcome.String())
+	recorder.AnnotatedEventf(app, outcomeEvent.Annotations, v1.EventTypeNormal, outcomeEvent.Reason, outcomeEvent.Description)
+	return nil
+}
+
+// appDeploymentEventFromWatchEvent converts a watch.Event into an AppDeploymentEvent
+func appDeploymentEventFromWatchEvent(watchEvent watch.Event, app *ketchv1.App) *AppDeploymentEvent {
+	event, ok := watchEvent.Object.(*v1.Event)
+	if !ok {
+		return nil
+	}
+	var version int
+	if len(app.Spec.Deployments) > 0 {
+		version = int(app.Spec.Deployments[len(app.Spec.Deployments)-1].Version)
+	}
+	return &AppDeploymentEvent{
+		Name:              app.Name,
+		DeploymentVersion: version,
+		Reason:            event.Reason,
+		Description:       event.Message,
+		Annotations: map[string]string{
+			DeploymentAnnotationAppName:            app.Name,
+			DeploymentAnnotationDevelopmentVersion: strconv.Itoa(version),
+			DeploymentAnnotationEventName:          event.Reason,
+			DeploymentAnnotationDescription:        event.Message,
+		},
+		InvolvedObjectName:      event.InvolvedObject.Name,
+		InvolvedObjectFieldPath: event.InvolvedObject.FieldPath,
+		SourceHost:              event.Source.Host,
+		SourceComponent:         event.Source.Component,
+	}
+}
+
+// newAppDeploymentEvent creates a new AppDeploymentEvent, creating Annotations for use when emitting App Events.
+func newAppDeploymentEvent(app *ketchv1.App, reason, desc string) *AppDeploymentEvent {
+	var version int
+	if len(app.Spec.Deployments) > 0 {
+		version = int(app.Spec.Deployments[len(app.Spec.Deployments)-1].Version)
+	}
+	return &AppDeploymentEvent{
+		Name:              app.Name,
+		DeploymentVersion: version,
+		Reason:            reason,
+		Description:       desc,
+		Annotations: map[string]string{
+			DeploymentAnnotationAppName:            app.Name,
+			DeploymentAnnotationDevelopmentVersion: strconv.Itoa(version),
+			DeploymentAnnotationEventName:          reason,
+			DeploymentAnnotationDescription:        desc,
+		},
+	}
+}
+
+func (a *AppDeploymentEvent) String() string {
```

Comment:
do we need this function if we use annotations?

Reply:
Good idea. I removed this func and moved the string presentation logic into the corresponding shipa PR: https://github.com/shipa-corp/shipa/pull/1028